File: Algorithms/p386_Lexicographical_Numbers_2.cpp
class Solution {
public:
    vector<int> lexicalOrder(int n) {
        vector<int> res(n);
        int i = 0;
        for(int j=1;j<10&&j<=n;++j){
            lexicalOrderHelper(res,i,n,j);
        }
        return res;
    }
    void lexicalOrderHelper(vector<int>& res, int& i, int n, int m) {
        res[i++] = m;
        for(int j=0;j<10&&m*10+j<=n;++j){
            lexicalOrderHelper(res,i,n,m*10+j);
        }
    }
};

File: Algorithms/p386_Lexicographical_Numbers/p386_Lexicographical_Numbers_2.py
class Solution(object):
    def lexicalOrder(self, n):
        """
        :type n: int
        :rtype: List[int]
        """
        res = [None]*n
        i = [0]
        for j in range(1,min(10,n+1)):
            self.lexicalOrderHelper(res,i,n,j)
        return res
        
    def lexicalOrderHelper(self, res, i, n, m):
        res[i[0]] = m
        i[0] += 1
        m *= 10
        for j in range(0,10):
            if m+j>n:
                break
            self.lexicalOrderHelper(res,i,n,m+j)
